### PR TITLE
Fix large writes

### PR DESCRIPTION
**Pull request by vlovich**
This fixes write on Windows & write/writev on MacOS for very large writes (#1128). Still left todo is a cap'n'proto serialization
test of a large message to double-check but maybe that's not needed?

This is based on #1129 so the diff will be a bit messed up until that diff lands & this one is rebased.

### CODE REVIEW

**Comment by vlovich** [comment]
@kentonv do you have any thoughts on how to fix the compiler error in https://github.com/capnproto/capnproto/pull/1130/checks?check_run_id=1658987563 ?

All that's being done in that spot is a mild refactor but it's complaining about the for loop that's part of the expanded macro. Before:

```
 TEST(Serialize, FileDescriptors) {
#if _WIN32 || __ANDROID__
  // TODO(cleanup): Find the Windows temp directory? Seems overly difficult.
  char filename[] = "capnproto-serialize-test-XXXXXX";
#else
  char filename[] = "/tmp/capnproto-serialize-test-XXXXXX";
#endif
  kj::AutoCloseFd tmpfile(mkstemp(filename));
  char filename[] = KJ_TEMP_FILE_TEMPLATE("capnproto-serialize-test");
  kj::AutoCloseFd tmpfile = kj::test::mkstemp_auto_erased(filename);
   ASSERT_GE(tmpfile.get(), 0);
 
#if !_WIN32
  // Unlink the file so that it will be deleted on close.
  // (For win32, we already handled this is mkstemp().)
  EXPECT_EQ(0, unlink(filename));
#endif

   {                                                                                                 
     TestMessageBuilder builder(7);                                                                  
     initTestMessage(builder.initRoot<TestAllTypes>());                                              
     writeMessageToFd(tmpfile.get(), builder);                                                       
   }
```

After:
```
 TEST(Serialize, FileDescriptors) {
   char filename[] = KJ_TEMP_FILE_TEMPLATE("capnproto-serialize-test");
   kj::AutoCloseFd tmpfile = kj::test::mkstemp_auto_erased(filename);
   ASSERT_GE(tmpfile.get(), 0);

   {                       <--- compiler error on some platforms as it's assumed to be part of the for loop expanded above??                                                                            
     TestMessageBuilder builder(7);                                                                  
     initTestMessage(builder.initRoot<TestAllTypes>());                                              
     writeMessageToFd(tmpfile.get(), builder);                                                       
   }
```

File: c++/src/kj/test-util.c++ [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#include "debug.h"
+#include "test-util.h"
+#include <cstdlib>
```
Comment (kentonv) [inline c++/src/kj/test-util.c++:24]: In KJ we prefer the original C header names rather than C++'s weird wrappers.

File: c++/src/kj/io-test.c++ [hunk 2]
```diff
@@ -26,11 +26,75 @@
...
+  KJ_ASSERT(gb < ONE_GB_ALIGNED_PREFIXES.size());
+  strcpy(ptr + gb * ONE_GB, (ONE_GB_ALIGNED_PREFIXES.begin() + gb)->cStr());
+};
+
+static void clearPrefix(char* ptr, int gb) {
+  memset(ptr + gb * ONE_GB, 0, (ONE_GB_ALIGNED_PREFIXES.begin() + gb)->size());
+};
+#endif
+
+TEST(Io, WriteLarge) {
```
[the remark this answers is not in the document]
Comment (vlovich) [reply c++/src/kj/io-test.c++:58]: Is there an already existing preprocessor macro that protects running in release? If not I'll just #ifdef it out although that opens up this fix to potentially bitrot eventually.

File: c++/src/kj/test-util.h [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+// would require refactoring the mkstemp code a bit to take into account the types
+// needed to do this at runtime.
+#define KJ_TEMP_FILE_TEMPLATE(tpl) tpl "-XXXXXX"
+#else
+#define KJ_TEMP_FILE_TEMPLATE(tpl) "/tmp/" tpl "-XXXXXX"
+#endif
+
+namespace kj {
+namespace test {
+AutoCloseFd mkstemp_auto_erased(char *tpl);
```
[the remark this answers is not in the document]
Comment (vlovich) [reply c++/src/kj/test-util.h:38]: I am partial to customizing the filename so that when you encounter a thousands of temporary files, you can easily identify what's generating them. I've found this trick to be generally helpful in the past but open to removing it. Let me know.

File: c++/src/kj/miniposix.h [hunk 4]
```diff
@@ -91,8 +92,24 @@ inline int close(int fd) {
...
+// internally limited to INT_MAX.
+#define KJ_OS_MAX_IO INT_MAX
+#elif INTPTR_MAX >= INT64_MAX
+// Normal POSIX platforms are limited to SSIZE_MAX. However on 64-bit platforms there's no point
+// saturating any input I/O lengths - that would be 9 exabytes & insane to do in 1 I/O call.
+// Using SIZE_MAX ensures that the kj::min evaluation should be elided and the read/write/writev
+// calls should inlined directly to the libc function.
+#define KJ_OS_MAX_IO SIZE_MAX
+#else
+// On 32-bit platforms make sure that trying to do large I/O still works correctly (saturating &
```
[the remark this answers is not in the document]
Comment (vlovich) [reply c++/src/kj/miniposix.h:107]: If I recall correctly (it's been a while since I first wrote & tested everything - should have taken better notes clearly), this was for FreeBSD which returns EINVAL when the value is > SSIZE_MAX. On 32-bit platforms means writes > 2GB would fail even though you would expect to be able to write 4GB. Does that sound reasonable? I think I wrote this comment *after* our discussion in the original bug report so I would have incorporated the feedback that oversized writes in general would be impossible...

https://www.freebsd.org/cgi/man.cgi?write(2)

File: c++/src/kj/io-test.c++ (hunk 2 above)
[the remark this answers is not in the document]
Comment (kentonv) [reply c++/src/kj/io-test.c++:58]: Unfortunately, no, there isn't an existing #ifdef to make things run only during the extended release tests. We'd need to add that in `super_test.sh`.

File: c++/src/kj/test-util.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (kentonv) [reply c++/src/kj/test-util.h:38]: I'd agree with you, except that with these files we're unlinking them immediately, so they'll never be visible on the filesystem anyway...

File: c++/src/kj/miniposix.c++ [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+namespace miniposix {
+#if __APPLE__
+ssize_t writev(int fd, struct iovec* iov, size_t iovcnt) {
+  // Workaround for writev POSIX bug on Apple (filed as FB8934446) that prevents writes > 2GB.
+  // The process for doing this is:
+  //   1. We find the iov entry that causes us to exceed a 2GB write.
+  //   2. We truncate that entry to only have a length that would get us to 2GB
+  //   3. We do the actual write.
+  //   4. We restore the original length in the modified iovec entry before returning the result.
+  // This is a bit suboptimal for very large vectored I/O as the complexity becomes O(nm) where n is
```
[the remark this answers is not in the document]
Comment (vlovich) [reply c++/src/kj/miniposix.c++:34]: Fixed the comment.

File: c++/src/kj/test-util.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (vlovich) [reply c++/src/kj/test-util.h:38]: Fair point. Generally I assume a process will get killed at any random point (including after creating the file but before unlinking) but this is now a corner case not worth worrying about. I'll kill the template.

File: c++/src/kj/io-test.c++ [hunk 6]
```diff
@@ -26,11 +26,75 @@
...
+#include "test-util.h"
 #include <kj/compat/gtest.h>
 
 namespace kj {
 namespace {
 
+#if INTPTR_MAX == INT64_MAX
+static constexpr size_t ONE_GB = 1 * 1024 * 1024 * 1024;
+static constexpr size_t BUFFER_SIZE = 3 * ONE_GB;
+static constexpr kj::StringPtr ONE_GB_ALIGNED_PREFIXES_STORAGE[] = {
```
[the remark this answers is not in the document]
Comment (vlovich) [reply c++/src/kj/io-test.c++:38]: Will be updated when I upload the next revision.

File: c++/src/kj/io-test.c++ (hunk 2 above)
[the remark this answers is not in the document]
Comment (vlovich) [reply c++/src/kj/io-test.c++:80]: This is what I have. Can't upload as the tests right now don't build on some platforms for some reason (i.e. don't want to accidentally submit broken code). Posted above the error message if you have thoughts on how I can resolve the issue.

```
diff --git a/c++/CMakeLists.txt b/c++/CMakeLists.txt
index 1a2ea097..0ef12e67 100644
--- a/c++/CMakeLists.txt
+++ b/c++/CMakeLists.txt
@@ -26,6 +26,7 @@ set(INSTALL_TARGETS_DEFAULT_ARGS
 # Options ======================================================================
 
 option(BUILD_TESTING "Build unit tests and enable CTest 'check' target." ON)
+option(BUILD_RELEASE_TESTING "Build & run a more expensive set of tests that should be run less frequently. Only takes effect when BUILD_TESTING is ON" OFF)
 option(EXTERNAL_CAPNP "Use the system capnp binary, or the one specified in $CAPNP, instead of using the compiled one." OFF)
 option(CAPNP_LITE "Compile Cap'n Proto in 'lite mode', in which all reflection APIs (schema.h, dynamic.h, etc.) are not included. Produces a smaller library at the cost of features. All programs built against the library must be compiled with -DCAPNP_LITE. Requires EXTERNAL_CAPNP." OFF)
 
diff --git a/c++/src/kj/CMakeLists.txt b/c++/src/kj/CMakeLists.txt
index 8f0dbb3b..797c48f0 100644
--- a/c++/src/kj/CMakeLists.txt
+++ b/c++/src/kj/CMakeLists.txt
@@ -238,6 +238,9 @@ if(BUILD_TESTING)
   target_link_libraries(kj-tests kj-test kj)
   add_dependencies(check kj-tests)
   add_test(NAME kj-tests-run COMMAND kj-tests)
+  if(BUILD_RELEASE_TESTING)
+    target_compile_definitions(kj-tests PRIVATE -DKJ_TEST_LARGE_IO=1)
+  endif()
 
   if(NOT CAPNP_LITE)
     add_executable(kj-heavy-tests
diff --git a/c++/src/kj/io-test.c++ b/c++/src/kj/io-test.c++
index 7a16b3aa..1677cf2f 100644
--- a/c++/src/kj/io-test.c++
+++ b/c++/src/kj/io-test.c++
@@ -77,6 +77,7 @@ static void setPrefix(char* ptr, int gb) {
 };
 #endif
 
+#ifdef KJ_TEST_LARGE_IO
 TEST(Io, WriteLarge) {
   // Check that writing a single large write > 2GB succeeds. This is primarily to workaround an
   // Apple POSIX compliance issue (filed to Apple as FB8934446). This test only makes sense on
@@ -111,6 +112,7 @@ TEST(Io, WriteLarge) {
   }
 #endif
 }
+#endif  // KJ_TEST_LARGE_IO
 
 TEST(Io, WriteVec) {
   // Check that writing an array of arrays works even when some of the arrays are empty.  (This
@@ -139,6 +141,7 @@ TEST(Io, WriteVec) {
   EXPECT_STREQ("foobar", buf);
 }
 
+#ifdef KJ_TEST_LARGE_IO
 TEST(Io, WriteVecLarge) {
   // Check that writing an array of arrays works when one a single writev entry is > 2GB. This is
   // primarily to workaround an Apple POSIX compliance issue (filed to Apple as FB8934446). This
@@ -187,6 +190,7 @@ TEST(Io, WriteVecLarge) {
   }
 #endif
 }
+#endif // KJ_TEST_LARGE_IO
 
 KJ_TEST("stringify AutoCloseFd") {
   int fds[2];
diff --git a/super-test.sh b/super-test.sh
index 08d072e0..fe9bc02b 100755
--- a/super-test.sh
+++ b/super-test.sh
@@ -180,7 +180,7 @@ while [ $# -gt 0 ]; do
       rm -rf cmake-build
       mkdir cmake-build
       cd cmake-build
-      doit cmake -G "Unix Makefiles" ..
+      doit cmake -G "Unix Makefiles" -DBUILD_RELEASE_TESTING=ON ..
       doit make -j$PARALLEL check
       exit 0
       ;;

```

My thought is that rather than totally compiling out the tests as I do now, I should just somehow make them "no-ops" by doing an early return so that at least the code can't bitrot syntactically (unless the release stuff is tested in some automated system regularly that would flag it on a breakage?).

[the remark this answers is not in the document]
Comment (kentonv) [reply c++/src/kj/io-test.c++:80]: Release tests run in CI but only after merging to master, so I wouldn't worry too much about bitrotting. Though I wonder if the CI machines have enough RAM and disk space available to handle this...

I'd use the autotools build rather than cmake. Otherwise if this test fails it'll look like a cmake-specific problem and people will be confused. autotools is our primary build system. Plus with autotools, you can directly pass CXXFLAGS to `configure` so you don't need to update the build files at all.

File: c++/src/kj/miniposix.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (kentonv) [reply c++/src/kj/miniposix.h:107]: Hmm, is it actually feasible to allocate even a 2GB array on a 32-bit platform? I realize the virtual address space theoretically supports it but I would think there would be some practical limitations in the way, like kernel/user address space separation, fragmentation (especially with ASLR), etc.

[the remark this answers is not in the document]
Comment (vlovich) [reply c++/src/kj/miniposix.h:107]: Coming back to this, apparently on MacOS, a 32-bit process can apparently allocate the full 4GB: https://forums.freebsd.org/threads/how-much-memory-freebsd-amd64-can-allocate-to-an-i386-binary.62171/page-2

That being said, macOS 10.15 released in fall 2019 doesn't support 32-bit apps anymore so what you wrote is probably true of all remaining systems. I don't know what minimum version of macOS is targeted.

[the remark this answers is not in the document]
Comment (kentonv) [reply c++/src/kj/miniposix.h:107]: OK, so it's theoretically possible... but in practice, the application would really have to go out of its way to allocate one contiguous array over 2GB without being blocked by fragmentation, and it couldn't possibly allocate a second such array. Any app doing that should be aware that it is risking overflow and should be treading carefully. So I feel like this is not something we really need to contort our code to handle.

If we ignore that case, can the code be simpler? Can we avoid the need to override `write()` then? Or at least, can we put less stuff in the header? Maybe `KJ_OS_MAX_IO` could be kept private within `miniposix.c++`?

**Comment by kentonv** [comment]
Superseded by #2198.
